src/functional_test_runner/lib/lifecycle.ts

```diff
@@ -17,31 +17,34 @@
  * under the License.
  */
 
+type Listener = (...args: any[]) => Promise<void> | void;
+export type Lifecycle = ReturnType<typeof createLifecycle>;
+
 export function createLifecycle() {
   const listeners = {
-    beforeLoadTests: [],
-    beforeTests: [],
-    beforeTestSuite: [],
-    beforeEachTest: [],
-    afterTestSuite: [],
-    testFailure: [],
-    testHookFailure: [],
-    cleanup: [],
-    phaseStart: [],
-    phaseEnd: [],
+    beforeLoadTests: [] as Listener[],
+    beforeTests: [] as Listener[],
+    beforeTestSuite: [] as Listener[],
+    beforeEachTest: [] as Listener[],
+    afterTestSuite: [] as Listener[],
+    testFailure: [] as Listener[],
+    testHookFailure: [] as Listener[],
+    cleanup: [] as Listener[],
+    phaseStart: [] as Listener[],
+    phaseEnd: [] as Listener[],
   };
 
-  class Lifecycle {
-    on(name, fn) {
+  return {
+    on(name: keyof typeof listeners, fn: Listener) {
       if (!listeners[name]) {
         throw new TypeError(`invalid lifecycle event "${name}"`);
       }
 
       listeners[name].push(fn);
       return this;
-    }
+    },
 
-    async trigger(name, ...args) {
+    async trigger(name: keyof typeof listeners, ...args: any[]) {
```

Comment:
What benefit does that have?

Reply:
I had tried to define the listener type in the `listeners` object, so that the calls to trigger would validate the argument types, but after getting it working I realized it was a ton of code and didn't really provide any value so I just went for the any. This service is a pretty niche service and I don't think people are loosing much by having to do a little digging. I'm thinking about getting rid of the event emitter style interface though and replacing it with a service that has individual methods like `beforeLoadTests(fn: (suite: Suite) => Promise<void> | void)`